fs/fuse/fuse_dlm_cache.c

```diff
@@ -0,0 +1,520 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * FUSE page lock cache implementation
+ */
+#include "fuse_i.h"
+#include "fuse_dlm_cache.h"
+
+#include <linux/list.h>
+#include <linux/types.h>
+#include <linux/slab.h>
+#include <linux/interval_tree_generic.h>
+
+/* Lock modes for FUSE page cache */
+#define FUSE_PCACHE_LK_READ 1 /* Shared read lock */
+#define FUSE_PCACHE_LK_WRITE 2 /* Exclusive write lock */
+
+/* Interval tree definitions for page ranges */
+static inline pgoff_t fuse_dlm_range_start(struct fuse_dlm_range *range)
+{
+	return range->start;
+}
+
+static inline pgoff_t fuse_dlm_range_last(struct fuse_dlm_range *range)
+{
+	return range->end;
+}
+
+INTERVAL_TREE_DEFINE(struct fuse_dlm_range, rb, pgoff_t, __subtree_end,
+		     fuse_dlm_range_start, fuse_dlm_range_last, static,
+		     fuse_page_it);
+
+/**
+ * fuse_page_cache_init - Initialize a page cache lock manager
+ * @cache: The cache to initialize
+ *
+ * Initialize a page cache lock manager for a FUSE inode.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+int fuse_dlm_cache_init(struct fuse_dlm_cache *cache)
+{
+	if (!cache)
+		return -EINVAL;
+
+	init_rwsem(&cache->lock);
+	cache->ranges = RB_ROOT_CACHED;
+
+	return 0;
+}
+
+/**
+ * fuse_page_cache_destroy - Clean up a page cache lock manager
+ * @cache: The cache to clean up
+ *
+ * Release all locks and free all resources associated with the cache.
+ */
+void fuse_dlm_cache_destroy(struct fuse_dlm_cache *cache)
+{
+	struct fuse_dlm_range *range;
+	struct rb_node *node;
+
+	if (!cache)
+		return;
+
+	/* Release all locks */
+	down_write(&cache->lock);
+	while ((node = rb_first_cached(&cache->ranges)) != NULL) {
+		range = rb_entry(node, struct fuse_dlm_range, rb);
+		fuse_page_it_remove(range, &cache->ranges);
+		kfree(range);
+	}
```

Comment:
We should probably rename it to fuse_dlm_cache_release_locks? And maybe pass in the fuse inode to make it clear it is per inode?